### PR TITLE
docs: [FC-0074] add docs about creating new filters with pipeline steps

### PR DESCRIPTION
**Pull request by mariajgrimaldi**
## Description

Update how-to documents to create and implement pipeline steps considering design suggestions introduced in #240.

Here are the main documents to review:
- How to Create a New Filter with Long-Term Support: https://github.com/openedx/openedx-filters/pull/242/files#diff-2825244e325626bcad68d204d59aa3e6ff696d1b5043fd4b990bd77e88dbb9a1
- Implement Pipeline Steps: https://github.com/openedx/openedx-filters/pull/242/files#diff-f1e167403e1377f568c3a2641fb3a151cd61d26887b3a468f91669d4494aee61

## Testing instructions

You can review the docs here: https://docsopenedxorg--242.org.readthedocs.build/projects/openedx-filters/en/242/how-tos/index.html

## Deadline

None

## Checklists

Check off if complete *or* not applicable:

**Merge Checklist:**
- [ ] All reviewers approved
- [ ] Reviewer tested the code following the testing instructions
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added with short description of the change and current date
- [ ] Documentation updated (not only docstrings)
- [ ] Code dependencies reviewed
- [ ] Fixup commits are squashed away
- [ ] Unit tests added/updated
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

**Post Merge:**
- [ ] Create a tag
- [ ] Create a release on GitHub
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] Upgrade the package in the Open edX platform requirements (if applicable)

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @mariajgrimaldi!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/hooks-extension-framework`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

File: docs/how-tos/create-a-pipeline-step.rst [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+First, add the ``openedx-filters`` plugin into your dependencies so the library's environment recognizes the event you want to consume. You can install ``openedx-filters`` by running:
+
+.. code-block:: bash
+
+   pip install openedx-filters
+
+This will mainly make the filters available for your CI/CD pipeline and local development environment. If you are using the Open edX platform, the library should be already be installed in the environment so no need to install it.
+
+Step 4: Create a Pipeline Step
```
Comment (sarina) [inline docs/how-tos/create-a-pipeline-step.rst:41]: This goes from step 2 to step 4!

File: docs/how-tos/create-a-pipeline-step.rst [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+- The method signature matches the filter's signature, accepting the user, course key, and enrollment mode as arguments and returning the same arguments if the email address is valid. You can also return an empty dictionary if you don't need to modify the data.
+
+Consider the following when creating a pipeline step:
+
+- Limit each step to a single responsibility to make the code easier to maintain and test.
+- Keep the pipeline step logic simple and focused on the specific task it needs to perform.
+- Consider the performance implications of the pipeline step and avoid adding unnecessary complexity or overhead, considering the pipeline will be executed each time the filter is triggered.
+- Implement error handling and logging in the pipeline step to handle exceptions and provide useful information for debugging, considering both development and production environments.
```
Comment (sarina) [inline docs/how-tos/create-a-pipeline-step.rst:76]: Does it make sense to include an example of logging or verbose error messaging in your simple example above?

File: docs/how-tos/create-a-pipeline-step.rst [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+   }
+
+Step 6: Test the Pipeline Step
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After creating the pipeline step and configuring the pipeline for the filter, you should test the pipeline step to ensure it works as expected. You can trigger the filter in your development environment and verify that the pipeline step is executed correctly. You should test different scenarios, including valid and invalid email addresses, to ensure the pipeline step behaves as expected.
+
+You should also implement unit tests for the pipeline step to verify its functionality and handle edge cases. Unit tests can help you identify issues early in the development process and ensure the pipeline step works as intended. To implement the unit test you can directly call the ``run_filter`` method of the filter definition and assert that the pipeline step behaves as expected. Or you can directly call the pipeline step class and assert that the method returns the expected output.
+
+In our example, you could write a unit test for the pipeline step like this:
```
Comment (sarina) [inline docs/how-tos/create-a-pipeline-step.rst:103]: Where should the unit tests live? (is it obvious or is it worth explaining?)

File: docs/reference/architecture-subdomains.rst [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+Currently, these are the `architecture subdomains`_ used by the Open edX Events library:
+
++-------------------+----------------------------------------------------------------------------------------------------+
+| Subdomain name    | Description                                                                                        |
++===================+====================================================================================================+
+| Content Authoring | Allows educators to create, modify, package, annotate (tag), and share learning content.           |
++-------------------+----------------------------------------------------------------------------------------------------+
+| Learning          | Allows learners to consume content and perform actions in a learning activity on the platform.     |
++-------------------+----------------------------------------------------------------------------------------------------+
+| Analytics         | Provides insights into learner behavior and course performance.                                    |
```
Comment (sarina) [inline docs/reference/architecture-subdomains.rst:13]:
```suggestion
| Analytics         | A look into learner behavior and course performance.                                    |
```

I don't love this but I think you should avoid using "insights" to not confuse with Insights, the deprecated analytics product.

**Review by sarina — APPROVED** [review]

File: docs/how-tos/create-a-pipeline-step.rst (hunk 1 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-pipeline-step.rst]: I totally missed this. Thank you!

File: docs/how-tos/create-a-pipeline-step.rst (hunk 2 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-pipeline-step.rst]: I added some. This suggestion mainly focuses on helping developers understand when the filter is doing what it's supposed to do based on the given conditions. Do you think the example illustrates that?

File: docs/how-tos/create-a-pipeline-step.rst (hunk 3 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-pipeline-step.rst:106]: Totally! I added the path where the tests should be located.

File: docs/reference/architecture-subdomains.rst (hunk 4 above)
Comment (mariajgrimaldi) [reply docs/reference/architecture-subdomains.rst]: I used "visbility" instead of insights, what do you think?

**Review by sarina — APPROVED** [review]

File: docs/how-tos/create-a-pipeline-step.rst [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+Assumptions
+-----------
+
+- You have a development environment set up using `Tutor`_.
+- You have a basic understanding of Python and Django.
+- You understand the concept of filters or have reviewed the relevant :doc:`/concepts/index` docs.
+- You are familiar with the terminology used in the project, such as the terms :term:`Filter Type`. If not, you can review the :doc:`../reference/glossary` docs.
+- You have reviewed the :doc:`../decisions/0007-filter-design-practices` ADR.
+- You have identified that you need to create a new filter and have a use case for the filter.
```
Comment (MaferMazu) [inline docs/how-tos/create-a-pipeline-step.rst:16]: My only comment is that we may want to drop the last assumption because, in this doc, we are talking about pipelines, and you cover the understanding part in Step 1.

**Review by MaferMazu — APPROVED** [review]
I have one suggestion, but the rest looks good to me.

Thanks @mariajgrimaldi ✨

File: docs/how-tos/create-a-pipeline-step.rst [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+-----
+
+To create a pipeline step for a filter in the Open edX platform, follow these steps:
+
+Step 1: Understand your Use Case and Identify the Filter to Use
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before creating a pipeline step, you should understand your use case for the filter and the specific logic you want to implement in the pipeline step. In our example, we want to prevent users from enrolling in a course if they do not have a valid email address. We will create a pipeline step that checks if the user's email address is valid and raise an exception if it is not.
+
+You should review the list of filters available in the Open edX platform and identify the filter that best fits your use case. In our example, we will use the `CourseEnrollmentStarted filter`_ to implement the logic for our use case. You should review the filter's arguments to understand the data that will be passed to the pipeline step and the expected output. This will help you define the pipeline step's logic and signature.
```
Comment (MaferMazu) [inline docs/how-tos/create-a-pipeline-step.rst:28]:
```suggestion
You should review the `list of filters available`_ in the Open edX platform and identify the filter that best fits your use case. In our example, we will use the `CourseEnrollmentStarted filter`_ to implement the logic for our use case. You should review the filter's arguments to understand the data that will be passed to the pipeline step and the expected output. This will help you define the pipeline step's logic and signature.
```

I added this suggestion because I saw something similar in the openedx-events PR, and it took me to Reference/Events. I didn't know that page existed. I think it should be helpful to be able to redirect the people to the list of filters here: https://docs.openedx.org/projects/openedx-filters/en/latest/reference/filters.html#

Note: In the suggestion, we need to reference the link because I didn't put it.

File: docs/how-tos/create-a-pipeline-step.rst (hunk 5 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-pipeline-step.rst]: Thanks for the suggestion! I replaced this line with understanding the pipeline step logic to implement instead: https://github.com/openedx/openedx-filters/pull/242/commits/667abe5c079d70fe34b9369c6b1880fefa54c6d3

File: docs/how-tos/create-a-pipeline-step.rst (hunk 6 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-pipeline-step.rst]: This is pretty useful, thank you! https://github.com/openedx/openedx-filters/pull/242/commits/ad4a7946180a54a14424f5ff8b4a63e829cf6d0b

**Comment by MaferMazu** [comment]
Thanks for addressing my feedback. This looks good to me ✅

File: docs/how-tos/create-a-new-filter.rst [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+- `Add Extensibility Mechanism to IDV to Enable Integration of New IDV Vendor Persona`_
+
+.. note:: If your use case is too specific to your organization, you can implement them in your own library and use it within your services by adopting an organization-scoped approach. However, if you think that your use case could be beneficial to the community, you should propose it to the community for feedback and collaboration.
+
+In our example our use case proposal could be:
+
+   I want to add a filter that will be triggered when a user enrolls in a course from the course about page. This filter will be used to prfilter users from enrolling in a course if they do not meet the eligibility criteria. The filter will be triggered when the user clicks the enroll button on the course about page and will check if the user meets the eligibility criteria. If the user does not meet the criteria, the filter will raise an exception to prfilter the user from enrolling in the course.
```
Comment (felipemontoya) [inline docs/how-tos/create-a-new-filter.rst:52]:
```suggestion
   I want to add a filter that will be triggered when a user enrolls in a course from the course about page. This filter will be used to prevent users from enrolling in a course if they do not meet the eligibility criteria. The filter will be triggered when the user clicks the enroll button on the course about page and will check if the user meets the eligibility criteria. If the user does not meet the criteria, the filter will raise an exception to prevent the user from enrolling in the course.
```

File: docs/how-tos/create-a-new-filter.rst [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+.. note:: Try placing the filter so it can be triggered before the process is completed, so it can alter the process if needed. In some cases, this would be at the beginning of the process, while in others it would be elsewhere.
+
+Step 7: Implement Your Pipeline Steps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Implementing pipeline steps allows you to modify the behavior of the application when the filter is triggered. Pipeline steps are a sequence of steps that are executed in a specific order to modify the behavior of the application. You can configure them with the :term:`filter configuration` to define the sequence of steps that are executed when the filter is triggered. Follow the steps in the :doc:`../how-tos/create-a-pipeline-step` guide to implement the pipeline steps for the filter.
```
Comment (felipemontoya) [inline docs/how-tos/create-a-new-filter.rst:156]: I think we should say here that the pipeline steps are meant to be written in the plugin code and not in the openedx-filters library.

File: docs/how-tos/create-a-pipeline-step.rst [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before creating a pipeline step, you should understand your use case for the filter and the specific logic you want to implement in the pipeline step. In our example, we want to prevent users from enrolling in a course if they do not have a valid email address. We will create a pipeline step that checks if the user's email address is valid and raise an exception if it is not.
+
+You should review the :doc:`list of filters <../reference/filters>` available in the Open edX platform and identify the filter that best fits your use case. In our example, we will use the `CourseEnrollmentStarted filter`_ to implement the logic for our use case. You should review the filter's arguments to understand the data that will be passed to the pipeline step and the expected output. This will help you define the pipeline step's logic and signature.
+
+Step 2: Install Open edX Filters
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, add the ``openedx-filters`` plugin into your dependencies so the library's environment recognizes the event you want to consume. You can install ``openedx-filters`` by running:
```
Comment (felipemontoya) [inline docs/how-tos/create-a-pipeline-step.rst:33]: ``openedx-filters`` plugin -> ``openedx-filters`` library. To not make it more confusing plugin should refer always to the developer's pluggable code and the openedx-filters or openedx-events should be called the library.

Comment (felipemontoya) [reply docs/how-tos/create-a-pipeline-step.rst:33]: "so the library's environment recognizes the event you want to consume" this is also confusing. We are not consuming an event, but creating a filter pipeline in this guide.

**Review by felipemontoya — CHANGES_REQUESTED** [review]
Some inline comments. Overall looks good, but I'm starting to see something that we are doing rather unintentionally. By focusing our documentation effort on all the parts of the lifecycle of : creating a new filter that did not previously exist, add said filter to the platform, implement pipeline steps with your feature, we are creating the impression on devs that they all need to take this route. I think we need to make it very well known that there are already quite a few filters and events available and that for every one of those you only need the part where you develop your code. This misconception is pushing people to believe that using filters is too complicated and not worth doing.

**Comment by mariajgrimaldi** [comment]
> Overall looks good, but I'm starting to see something that we are doing rather unintentionally. By focusing our documentation effort on all the parts of the lifecycle of : creating a new filter that did not previously exist, add said filter to the platform, implement pipeline steps with your feature, we are creating the impression on devs that they all need to take this route. I think we need to make it very well known that there are already quite a few filters and events available and that for every one of those you only need the part where you develop your code.

I hear what you're saying, but I’m not sure I follow, given that one of the first steps in creating a pipeline specifically mentions reviewing the list of available filters to identify the right one for the use case: https://docsopenedxorg--242.org.readthedocs.build/projects/openedx-filters/en/242/how-tos/create-a-pipeline-step.html#step-1-understand-your-use-case-and-identify-the-filter-to-use. Do you think that part isn’t clear enough? I’m open to suggestions to make it more obvious.

File: docs/how-tos/create-a-new-filter.rst (hunk 7 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-new-filter.rst:52]: Thank you for noticing!

File: docs/how-tos/create-a-new-filter.rst (hunk 8 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-new-filter.rst:156]: There is a note about it here: https://github.com/openedx/openedx-filters/pull/242/files#diff-f1e167403e1377f568c3a2641fb3a151cd61d26887b3a468f91669d4494aee61R57. But I agree adding it here could make it clearer.

File: docs/how-tos/create-a-pipeline-step.rst (hunk 9 above)
Comment (mariajgrimaldi) [reply docs/how-tos/create-a-pipeline-step.rst:33]: I borrowed this from the consume events docs, that's why it is all mixed up. Thank you for noticing! Although I now have a question, technically openedx-events _is_ a plugin, do you think that would cause misunderstandings? Calling openedx-events a plugin instead?

**Comment by mariajgrimaldi** [comment]
@felipemontoya: I've addressed all your comments in this commit https://github.com/openedx/openedx-filters/pull/242/commits/2ee41e3af51e863fee70b0647c56c15cf899b56d. Thank you again!

File: docs/how-tos/create-a-pipeline-step.rst (hunk 9 above)
Comment (felipemontoya) [reply docs/how-tos/create-a-pipeline-step.rst]: What threshold do you use to recognize something as a plugin? For me the defining characteristic of a plugin, is that it defines the `plugin_app` in its apps.py config. Neither events nor filters do that. That is why I call them libraries.

I could certainly be picking the wrong definition here. The other approach that I use for this, which is what triggered the comment, is that plugin is something that is not part of the core. We try to gently push developers in the direction of using plugins for the  custom development instead of forking. By this approach also events and filters would not be considered plugins.

Comment (mariajgrimaldi) [reply docs/how-tos/create-a-pipeline-step.rst]: Sorry for the misunderstanding. opened-events _is_ a Django application which makes it more than a library, and I thought it actually defined a ``plugin_app``, but after reviewing [apps.py](https://github.com/openedx/openedx-events/commit/29fc69e955b6bed941f069ab355ae63f07e37d2a) I realized it didn't.

**Comment by sarina** [comment]
@mariajgrimaldi once again moving filenames has broken the docs.openedx.org build: https://readthedocs.org/projects/docsopenedxorg/builds/27044031/

Could you please either refrain from moving filenames further, or make a dummy pr to docs.openedx.org when moving filenames to ensure you're not breaking our build over there?

**Comment by mariajgrimaldi** [comment]
@sarina: I'm sorry for breaking the build again. I missed it by accident.

This was the last docs PR that changed file names so I'm hoping this won't happen again.

I'll work on this first time tomorrow.
